### ISSUE TITLE
FileSystemWatcher.OSX consumes way too many threads

### ISSUE THREAD

**Issue by Therzok** [dotnet/corefx#30600]
Given a user of the FSW which wants to monitor only changes on some files, the implementation of the OSX watcher, starves the system by creating one thread per watcher.

Without writing a helper class that limits the number of watchers by doing some heuristics on the path trees that need to be watched and manually bookkeeping the watchers, it is quite impossible to workaround this, so it seems like everyone ends up having to write glue code / their own watcher implementation.

Few things to note:

* The same FSEventStream could be reused between multiple FSWs.

For the FSEventStream, there is a [pathRoots](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs#L236) argument that always uses [1 path](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs#L217).

* A thread can run multiple CFRunLoops

Given any thread, you can run multiple CFRunLoops serially, the same as one is run, using CFRunLoopRunInMode

* The Inotify watcher seems to use a _task_ per watcher, not a _thread_, but that does not seem a complete solution.

**Comment by Therzok** [comment]
Another thing that comes to mind is to have one CFRunLoop shared by all the FSEventStreams

**Comment by marek-safar** [comment]
@karelz who does own this area and can provide input on prefered implementation fixes?

**Comment by stephentoub** [comment]
> who does own this area

https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md lists area owners.
cc: @JeremyKuhne 

> can provide input on prefered implementation fixes?

It'd be great if you could propose a solution / provide a PR to address the scalability issues you're hitting. If there's a good way to do it by leveraging a single thread to run multiple CFRunLoops, that sounds like the best approach.

**Comment by marek-safar** [comment]
Copying proposals from different issue

>Other solutions would be to make one thread run all the runloops for filesystemwatchers or have some kind of scheduler with more than 1 thread.

>Or some way to use the FSEventStream overload which takes multiple root directories to watch.

**Comment by Therzok** [comment]
# ❓ IDE Alternative
There's also a different solution that we have used inside VSMac itself, but cannot be applied to all extensions integrating with the IDE:
```
That is to use a PathTree structure that is modified every time a new FileSystemWatcher is requested on a path.

The tree emulates a filesystem tree into an actual in-memory tree. Every path sent in is processed and the following happens: every directory becomes a node, every path separator is an edge. Each node also tracks liveness, so we know whether a given directory is a node requested as a watcher directory or if it's been created just to be able to create the whole tree.

Then, the PathTree itself provides a really efficient algorithm to limit the number of directories to watch. The Normalize method will use a queue and a maximum number of nodes to watch.

The normalize method is a BFS that yields live nodes until the number of yielded nodes is smaller than the maximum one, or yields parent nodes when we would exceed that limit.

https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Core/MonoDevelop.FSW/PathTree.cs
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Core/MonoDevelop.FSW/PathTreeNode.cs
```

This has the drawback of being done only in the application level, not being able to prevent extensions from using the native APIs and creating too many threads.

Also, the system itself reconstructs the filewatchers based on how the algorithm decides to change the paths to watch, so that could prevent some information from being passed on.

# ❌ : Share Thread for CFRunloops
I've further read into apple documentation, it seems like a runloop puts a thread to sleep when there is no event to process.

# ✅ Ideas - Share CFRunLoop for FSEventStreams
The immediate solution I see viable is sharing the CFRunLoop between the FSEventStream's, via passing the same NSThread runloop to all the FSEventStream scheduling calls.

The problem then becomes dispatching the events to the right FileSystemWatcher instances.

When this is done, we would probably have duplicate events being processed on the same runloop, as one directory could be monitored by 2 different FSEventStream sources. So events would probably end up being triggered twice for the same FileSystemWatcher. Therefore de-duplication of registrations or a way to map an event to a filesystemwatcher should be done.

**Comment by JeremyKuhne** [comment]
@Therzok I think that sharing the runloop is worth attempting.

**Comment by Therzok** [comment]
I suppose this can be closed due to #33865
